docs/deployment-design.md

```diff
@@ -0,0 +1,57 @@
+
+* Our on-prem server only has enough resources to run one model at a time, so only one data scientist may be running modeling experiments at a given time
+  * Further, our server has no notion of a job queue, so a data scientist who is waiting to run a model must notice that a previous run has completed and initiate their run manually 
+* Our on-prem server does not have a GPU, so it can't make use of GPU-accelerated libraries like XGBoost
+* Model runs are decoupled from version control changes, so data scientists have to remember to commit their changes correctly
+* Results of model runs are not easily accessible to PR reviewers
+
+The design described below aims to remove these limitations while retaining as much simplicity, maintainability, and affordability as possible.
+
+## Requirements
+
+At a high level, a model deployment pipeline should:
+
+* Integrate with our existing cloud infrastructure (GitHub and AWS)
+* Trigger model runs from pull request branches
+* Require code authors to approve model runs before they are initiated
+* Run the model on ephemeral, cheap, and isolated cloud infrastructure
+* Run multiple model runs simultaneously on separate hardware
+* Report model statistics back to the pull request that triggered a run
+
+## Design
+
+Here is a rough sketch of a new model deployment pipeline:
+
+* Define a new workflow, `run-model.yaml`, that runs on:
+  * Every commit to every pull request against the main branch
+  * The `workflow_dispatch` event
+* Set up the workflow so that it deploys to the `staging` environment and requires [manual approval](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#using-environments-to-manually-trigger-workflow-jobs)
+* Use the [`configure-aws-credentials`](https://github.com/aws-actions/configure-aws-credentials) action to authenticate with AWS
+* Run Terraform to make sure an AWS Batch job queue and job definition exist for the PR
+* Build and push a new Docker image to ECR
+* Use the AWS CLI to [submit a job](https://docs.aws.amazon.com/cli/latest/reference/batch/submit-job.html) to the Batch queue
+* Use the AWS CLI to [poll the job status](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/batch/describe-jobs.html) until it has a terminal status (`SUCCEEDED` or `FAILED`)
+  * Once the job has at least a `RUNNING` status, use the `jobStreamName` parameter to print a link to its logs
+* _TK: How should we format output for pushing back to the PR? Is it enough to push the results doc to S3 and print a link in the workflow output, or do we need an integration that can actually post a comment on the PR, à la Codecov?_
+
+## Tasks
+
+We will create GitHub issues for the following tasks:
+
+* Add Docker image definition for the model
```

Comment:
Note that there's an [old GitLab CI script](https://gitlab.com/ccao-data-science---modeling/models/ccao_res_avm/-/blob/master/.gitlab-ci.yml?ref_type=heads) that's basically a prototype of the pipeline laid out here. It might be useful.

Reply:
This is helpful! I added a pointer to it in c0ffa5a.